Feature/gemini object detection

onuralpszr opened the pull request:
## 🚀 from_vlm now has Google gemini 2D  spatial understanding support  for Detection class 🎯


[![Open in Colab](https://img.shields.io/badge/Open%20in-Colab-blue?logo=googlecolab)](https://colab.research.google.com/drive/1DVNI1T5i4JzzNokSTjj9RBKqWOUNvim5?usp=sharing)

## ✨ New Functionality:

* [`supervision/detection/core.py`](diffhunk://#diff-d620bf6569b39b0318ed01ac9fbd645b801ee1e237c188e3c50a7e3480f2ee25R40): Added support for `from_google_gemini` and included an example in the `from_lmm` method documentation. [[1]](diffhunk://#diff-d620bf6569b39b0318ed01ac9fbd645b801ee1e237c188e3c50a7e3480f2ee25R40) [[2]](diffhunk://#diff-d620bf6569b39b0318ed01ac9fbd645b801ee1e237c188e3c50a7e3480f2ee25L718-R719) [[3]](diffhunk://#diff-d620bf6569b39b0318ed01ac9fbd645b801ee1e237c188e3c50a7e3480f2ee25L846-R901) [[4]](diffhunk://#diff-d620bf6569b39b0318ed01ac9fbd645b801ee1e237c188e3c50a7e3480f2ee25R954-R958)
* [`supervision/detection/vlm.py`](diffhunk://#diff-c4a61ae6e74887622a7248a14d9bc5dcb301ec23e8920b70578e7105fa82df5dR20-R48): Added `GOOGLE_GEMINI_2_0` to `LMM` and `VLM` enums, and implemented the `from_google_gemini` function. [[1]](diffhunk://#diff-c4a61ae6e74887622a7248a14d9bc5dcb301ec23e8920b70578e7105fa82df5dR20-R48) [[2]](diffhunk://#diff-c4a61ae6e74887622a7248a14d9bc5dcb301ec23e8920b70578e7105fa82df5dL235-R242) [[3]](diffhunk://#diff-c4a61ae6e74887622a7248a14d9bc5dcb301ec23e8920b70578e7105fa82df5dL297-R314) [[4]](diffhunk://#diff-c4a61ae6e74887622a7248a14d9bc5dcb301ec23e8920b70578e7105fa82df5dR324-R387)

## 🔧 Pre-commit update


* [`.pre-commit-config.yaml`](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9L28-R35): Updated `bandit` to `1.8.3`, `ruff-pre-commit` to `v0.9.6`, and `codespell` to `v2.4.1`. [[1]](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9L28-R35) [[2]](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9L51-R51)

Assertion formatting improvements:
* Various test files: Reformatted assertions for better readability and consistency. [[1]](diffhunk://#diff-2185113c87138ca76c082af73589d45a6d3be64a4967cebeb1d07516ba07281aL411-R413) [[2]](diffhunk://#diff-a50b39fd7f1df19264075af2e2d41dfcb520566b1ef01f5b664be9ee4bae743dL245-R247) [[3]](diffhunk://#diff-106334deba835d12e663ee7c0961c5eb4f4215a364e3a847af123ed95aa430faL253-R258) [[4]](diffhunk://#diff-0379a32ca37a7f402d04f932db914ea57a505705c31a01ea57d483524bc374e3L392-R398) [[5]](diffhunk://#diff-0379a32ca37a7f402d04f932db914ea57a505705c31a01ea57d483524bc374e3L1045-R1051) [[6]](diffhunk://#diff-0379a32ca37a7f402d04f932db914ea57a505705c31a01ea57d483524bc374e3L1218-R1224) [[7]](diffhunk://#diff-788347cdcfb7a14d2fcfbe29847050abab7f285bac72be83471f18e2515c12d4L183-R185)

onuralpszr commented:
cc @SkalskiP  FRIENDLY PING for review please

soumik12345 commented:
Hi @onuralpszr, I was unable to reproduce the results using the attached colab, the bounding boxes don't look correct.

Result with gemini-2.0-flash 👇
![image](https://github.com/user-attachments/assets/2f24224f-dc10-4e3f-bcee-bac5a9f7207c)

Result with gemini-2.5-flash 👇
![image](https://github.com/user-attachments/assets/9664dc63-43be-4789-874d-fa9b25442e13)

Result with gemini-2.5-pro 👇
![image](https://github.com/user-attachments/assets/d7d6adef-8707-4eb0-9f1b-6a4897b50039)

onuralpszr commented:
cc @SkalskiP @soumik12345 fixes are added and new results pictures are also added. I also updated colab for testing multiple different gemini model easily

soumik12345 reviewed (approved):
LGTM!

SkalskiP commented on supervision/detection/vlm.py:
```diff
@@ -315,3 +350,64 @@ def from_florence_2(
+    resolution_wh: Tuple[int, int],
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Parse and scale bounding boxes from Google Gemini style JSON output.
+    https://aistudio.google.com/
+    https://ai.google.dev/gemini-api/docs/vision?lang=python
```
replace this with:

Parse and scale bounding boxes from Google Gemini style [JSON output](https://ai.google.dev/gemini-api/docs/vision?lang=python).

SkalskiP commented on supervision/detection/utils.py:
```diff
@@ -447,6 +447,62 @@ def xyxy_to_xcycarh(xyxy: np.ndarray) -> np.ndarray:
     return result.astype(float)
 
 
+def normalized_xyxy_to_absolute_xyxy(
```
please rename this function to `denormalize_boxes` to match existing `clip_boxes`, `pad_boxes` and `scale_boxes`

SkalskiP commented on supervision/detection/core.py:
````diff
@@ -843,12 +844,65 @@ def from_lmm(
             # array([0])
             ```
-        """
+
````
include `qwen_2_5_vl` example as well

SkalskiP commented on supervision/detection/vlm.py:
```diff
@@ -315,3 +350,64 @@ def from_florence_2(
     assert False, f"Unimplemented task: {task}"
+
+
+def from_google_gemini(
+    result: str,
+    resolution_wh: Tuple[int, int],
+) -> Tuple[np.ndarray, np.ndarray]:
```
This API is inconsistent with the `from_paligemma` and `from_qwen_2_5_vl` implementations.

Currently, it is not possible to resolve `class_id` values. To address this, we should allow users to optionally provide a `classes: Optional[List[str]] = None` argument. If this argument is given, we should attempt to resolve the `class_id` for each detection, following the same approach as in `from_paligemma` and `from_qwen_2_5_vl`.

The function should return `Tuple[np.ndarray, Optional[np.ndarray], np.ndarray]`, consistent with the return type of `from_paligemma` and `from_qwen_2_5_vl`, where `Optional[np.ndarray]` corresponds to `class_id`. This value should be `None` if `classes` is not provided, and an `np.ndarray` if it is.
